### PR TITLE
tenantcostserver: update token bucket debt algorithm

### PR DESCRIPTION
**Pull request by andy-kimball**
Previously, the token bucket's debt algorithm was greedy, always granting as many tokens as available on a first-come-first-served basis. This resulted in inequitable treatment of competing tenant instances, with some instances being starved while others got more than their share. Depending on the timing of calls, the balance could rapidly shift, with big swings in the relative RU rates of tenant instances.

This PR updates the algorithm to track the average debt in the bucket. Grants to requesting tenant instances are limited to the difference between the limit of tokens in the bucket and the average debt. This means that requests that come in when there are many tokens available will not greedily take all the tokens; instead they'll get a portion that corresponds to the average number of tokens that are available at any given time.

The new algorithm incrementally moves towards "fairness" across tenant instances, even in the presence of shifting workloads or tenant instances joining/leaving.

Epic: CC-25942

Release note: None

### CODE REVIEW

**Review by jeffswenson — APPROVED** [review]
LGTM

**Review by stevendanna — COMMENTED** [review]
Cool to see us get more sophisticated here. I'm interested to see where this heads over time.

I haven't thought very long about the strategy here. I assume this has been and will continue to be tuned based on observations in production. For example, I wonder a little about whether the `movingAvgFactor` used will provide the right level of smoothing given that our unit of "time" is a request. But I don't have very much intuition here and defer to y'alls judgement.

Is the goal with re-using the currently unused column to be able to backport this change or the convenience of not having to do a system schema change?

Minor procedural note: I wonder if we really need the MT team as "codeowners" of these files.  I'm always happy to take a look as I find y'alls work interesting, but it seems like there might be better group to _require_ review from.

**Review by andy-kimball — COMMENTED** [review]
I've done quite a bit of prototyping and testing of this algorithm. While I'm sure it could be improved, it does pretty well in a variety of different situations (multiple tenant instances, each with different loads that turn on/off unpredictably).

RE: the unused column, we do need to backport this to 23.2, under the heading of "new feature work with high business impact". All of the changed code only runs in Serverless clusters, so there should be no impact elsewhere.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112909#-:-NhbSB-c2YbUtJB8doJv:b-el6ltj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson)

___
*[`pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/tenant_token_bucket.go` line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/112909#gh-1370125339:-NhYo9G92xzR2zmzf19V:bcazpjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4058e110600b52aafb9f17c49f485f249baf483f/pkg/ccl/multitenantccl/tenantcostserver/tenanttokenbucket/tenant_token_bucket.go#L52)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Re-using system.tenant_usage.current_share_sum feels risky from a mixed version standpoint. Typically a behavior change like this would be attached to a version gate. But reading through the code it seems fine. When the kv servers are mixed version, the new servers will update the running average and the old servers will pass it through unmodified.

There is an assertion failing in tests, but it only runs in debug builds.
</blockquote></details>

None of this code will get meaningfully exercised in today's system, since we shut down the cluster once we reach 0 request units. `RUCurrentAvg` is only used once the token bucket goes into debt. So there should not be mixed version problems.

**Pull request merged**
